# 05_reranking: Dense Retrieval and Cross-Encoder Reranking

This notebook demonstrates the two-stage retrieval pipeline pattern:
1. **Bi-Encoder Stage**: Retrieves candidate passages using fast dense vector similarity lookups ($O(M)$ candidate matching).
2. **Cross-Encoder Stage**: Reranks the top candidates using a Cross-Encoder model, which performs full self-attention over the query and candidate document inputs simultaneously.

### Two-Stage Retrieval Math
- **Bi-Encoder**: Independently encodes query $q$ and passage $p$ into $\mathbf{u} = \text{enc}(q)$ and $\mathbf{v} = \text{enc}(p)$. Similarity score is the dot product $\mathbf{u} \cdot \mathbf{v}$. This is fast, as candidate matches can be computed via index space partitions (e.g. FAISS).
- **Cross-Encoder**: Feeds query and passage jointly to the Transformer architecture:
  $$s(q, p) = \text{Transformer}([q; \text{[SEP]}; p])$$
  This produces a unified score utilizing cross-attention weights between terms, making it computationally heavy ($O(N \cdot L^2)$ for $N$ candidates of length $L$) but extremely precise.

In [1]:
import os
from dotenv import load_dotenv
from langchain_community.vectorstores import FAISS
from langchain_openai import OpenAIEmbeddings
from sentence_transformers import CrossEncoder

load_dotenv(dotenv_path=r"d:\\Study\\Prep\\.env")

corpus = [
    "Gradient descent optimization requires setting an initial learning rate to update weights.",
    "Cross-entropy loss measures the performance of a classification model whose output is a probability value.",
    "Reranking uses a cross-encoder model to re-evaluate the relevance scores of a query and candidate documents.",
    "A cross-encoder processes query and document simultaneously to output a single score, yielding high accuracy.",
    "Bi-encoders embed queries and documents independently to allow rapid sub-millisecond retrieval via vector search.",
    "Regularization terms like L1 and L2 prevent neural network overfitting by penalizing large weights."
]

embeddings = OpenAIEmbeddings()
db = FAISS.from_texts(corpus, embeddings)
retriever = db.as_retriever(search_kwargs={"k": 5})

C:\Users\aryan\AppData\Local\Temp\ipykernel_31324\3568747537.py:3: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.vectorstores import FAISS


D:\Study\Prep\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
# Dense Candidate search
query = "How does cross-encoder reranking improve search accuracy?"
initial_docs = retriever.invoke(query)
print("Initial FAISS Dense Matches:")
for i, doc in enumerate(initial_docs):
    print(f"Rank {i+1}: {doc.page_content}")

Initial FAISS Dense Matches:
Rank 1: Reranking uses a cross-encoder model to re-evaluate the relevance scores of a query and candidate documents.
Rank 2: A cross-encoder processes query and document simultaneously to output a single score, yielding high accuracy.
Rank 3: Bi-encoders embed queries and documents independently to allow rapid sub-millisecond retrieval via vector search.
Rank 4: Cross-entropy loss measures the performance of a classification model whose output is a probability value.
Rank 5: Regularization terms like L1 and L2 prevent neural network overfitting by penalizing large weights.


In [3]:
# Cross-Encoder Reranking
model_name = "cross-encoder/ms-marco-MiniLM-L-6-v2"
try:
    reranker = CrossEncoder(model_name)
    pairs = [[query, doc.page_content] for doc in initial_docs]
    scores = reranker.predict(pairs)
    
    reranked_docs = sorted(zip(initial_docs, scores), key=lambda x: x[1], reverse=True)
    print("\nReranked Matches:")
    for i, (doc, score) in enumerate(reranked_docs):
         print(f"Rank {i+1} [CrossEncoder Score: {score:.4f}]: {doc.page_content}")
except Exception as e:
    print("CrossEncoder failed, using fallback ranking:", e)
    print("\nReranked Matches (Mocked):")
    print("Rank 1 [CrossEncoder Score: 0.8921]: Reranking uses a cross-encoder model to re-evaluate...")
    print("Rank 2 [CrossEncoder Score: 0.7410]: A cross-encoder processes query and document simultaneously...")

Loading weights:   0%|          | 0/105 [00:00<?, ?it/s]

Loading weights: 100%|██████████| 105/105 [00:00<00:00, 26251.90it/s]


Reranked Matches:
Rank 1 [CrossEncoder Score: 5.4267]: Reranking uses a cross-encoder model to re-evaluate the relevance scores of a query and candidate documents.
Rank 2 [CrossEncoder Score: 3.9401]: A cross-encoder processes query and document simultaneously to output a single score, yielding high accuracy.
Rank 3 [CrossEncoder Score: -3.3693]: Bi-encoders embed queries and documents independently to allow rapid sub-millisecond retrieval via vector search.
Rank 4 [CrossEncoder Score: -7.3123]: Cross-entropy loss measures the performance of a classification model whose output is a probability value.
Rank 5 [CrossEncoder Score: -11.2174]: Regularization terms like L1 and L2 prevent neural network overfitting by penalizing large weights.


### Output Explanation & Reranking Trace

#### Executed Results:
- **Bi-Encoder FAISS Matches**:
  - Rank 1: *'Reranking uses a cross-encoder model to re-evaluate...'*, Rank 2: *'A cross-encoder processes query...'*, Rank 3: *'Bi-encoders embed...'*, Rank 4: *'Cross-entropy loss...'*, Rank 5: *'Regularization terms...'*
- **Cross-Encoder Reranked Matches**:
  - Rank 1 [Score `5.4267`]: *'Reranking uses a cross-encoder model to re-evaluate the relevance scores of a query and candidate documents.'*
  - Rank 2 [Score `3.9401`]: *'A cross-encoder processes query and document simultaneously to output a single score, yielding high accuracy.'*
  - Rank 3 [Score `-3.3693`]: *'Bi-encoders embed queries and documents independently to allow rapid sub-millisecond retrieval via vector search.'*
  - Rank 4 [Score `-7.3123`]: *'Cross-entropy loss measures the performance of a classification model whose output is a probability value.'*
  - Rank 5 [Score `-11.2174`]: *'Regularization terms like L1 and L2 prevent neural network overfitting by penalizing large weights.'*

#### Detailed Logits Analysis:
- The Cross-Encoder model outputs raw classification logits indicating matching confidence.
- **Highly Relevant Candidates** (Rank 1 and Rank 2) received strong positive logit scores (`5.4267` and `3.9401`).
- **Conceptually Adjacent Candidates** (Rank 3, discussing Bi-encoders) was assigned a negative logit score (`-3.3693`), identifying it as secondary context.
- **Irrelevant Noise Candidates** containing overlap words (Rank 4 containing 'entropy/cross', and Rank 5 containing 'L1/L2') were strongly penalized by the model with scores of `-7.3123` and `-11.2174`.

This illustrates how stage-two Cross-Encoders resolve bi-encoder retrieval errors by modeling joint term-level cross-attention.